modules/saml/lib/IdP/SAML2.php

```diff
@@ -87,10 +90,21 @@ public static function sendResponse(array $state) {
 	 */
 	public static function handleAuthError(SimpleSAML_Error_Exception $exception, array $state) {
 		assert('isset($state["SPMetadata"])');
-		assert('isset($state["saml:ConsumerURL"])');
+		assert('array_key_exists("saml:ConsumerURL", $state)'); // Can be NULL.
 		assert('array_key_exists("saml:RequestId", $state)'); // Can be NULL.
 		assert('array_key_exists("saml:RelayState", $state)'); // Can be NULL.
 
+ 		if ($exception instanceof SimpleSAML_Error_Error && $exception->getErrorCode() === 'WRONGUSERPASS') {
+ 			/*
+ 			 * This should only happen during SOAP authentication, and indicates
+ 			 * that we do not have a valid username or password.
+ 			 */
+ 			header('WWW-Authenticate: Basic realm="IDPAUTH"', TRUE, 401);
+ 			header('Content-Type: text/plain');
+ 			echo('SOAP authentication requires basic auth.');
+ 			exit(0);
+ 		}
+
```

Comment:
Looking more into it, I think this whole if clause is intended to deal with missing authentication headers, but not with errors. That reaffirms my thoughts on the need for a specific code for the exception, different than *WRONGUSERPASS*.

I also think we need another if clause for `SimpleSAML_Error_Error` exceptions with a different code (say *ECP_AUTH_FAILURE*, for example), and act accordingly by sending back a SAML response to the ECP client informing about the error.